### ISSUE TITLE
DocumentSnapshot.ConvertTo<T> System.ArgumentException

### ISSUE THREAD

**Issue by DenSmoke** [GoogleCloudPlatform/google-cloud-dotnet#2426]
I get exception System.ArgumentException: "Unable to convert value type IntegerValue to System.Double" when trying to convert DocumentSnapshot to my attributed class.
I think this exception is excess because there is implicit conversion from integer to double in C#

**Comment by jskeet** [comment]
I don't want to start adding every possible implicit conversion in - and I believe it would be dangerous to do so. If you've got a field which is an integer in Firestore, but you try to use it as a `double` in your attributed class, that means when the value is serialized *back* to Firestore, it would overwrite the integer value with a floating point value, and anything which was assuming it was an integer would then have a problem.

Basically, I think you need to be consistent in your schema everywhere, otherwise you're creating problems for yourself. Using implicit conversions when deserializing would delay finding those problems, rather than actually solving them.

**Comment by DenSmoke** [comment]
Yeah but I don't want to have integer in Firestore, it supposed to be real number, but Firestore decises it for me because in some cases numbers are integer and I can't do anything with it

**Comment by jskeet** [comment]
> Yeah but I don't want to have integer in Firestore, it supposed to be real number, but Firestore decises it for me because in some cases numbers are integer and I can't do anything with it

What do you mean by "Firestore decides for me"? You should absolutely be able to store the type you want. I believe that storing the value with the type you want should be the focus of your efforts. What stored the integer value in the first place?

**Comment by DenSmoke** [comment]
Application on Android or iOS creates document in Firestore using value from textbox (stores "cost") parsed to double. Sometimes cost has no decimal point so this value becomes integer in Firestore

**Comment by jskeet** [comment]
> Sometimes cost has no decimal point so this value becomes integer in Firestore

I don't believe it should do that if you've actually parsed it to a `double`. For example, in the Java SDK's UserDataConverter class, [a double always ends up converted to DoubleValue](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java#L119)

I suggest you check the code you're using on the Android and iOS applications: check the behaviour of each of them separately, and look at the document that gets stored. Fixing the code that's writing the document to do the right thing is going to be more productive than dealing with different types, IMO.

**Comment by jskeet** [comment]
Great - thanks for letting us know.
